### ISSUE TITLE
File descriptor leak in every service

### ISSUE THREAD

**Issue by chaen** [DIRACGrid/DIRAC#3905]
TL;DR
====

We do not properly close the socket when we have an exception in the `read` or `write` call, leading in a socket leak, itself leading  to a complete crash of the service (file descriptor exhaustion). 

Details
=====

Under normal circumstances, the communication on the server side goes as follow (those are strace captures):

```
29286 accept(3, {sa_family=AF_INET6, sin6_port=htons(47950), inet_pton(AF_INET6, "::ffff:192.68.51.225", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 249
29286 getpeername(249, {sa_family=AF_INET6, sin6_port=htons(47950), inet_pton(AF_INET6, "::ffff:192.68.51.225", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 0
29286 fcntl(249, F_GETFL)               = 0x2 (flags O_RDWR)
29286 fcntl(249, F_SETFL, O_RDWR|O_NONBLOCK) = 0
23944 read(249,  <unfinished ...>
23944 read(249,  <unfinished ...>
23944 write(249, "\26\3\1\0:\2\0\0006\3\1\211I\2737\244\326\233\252?X\333\4\6*wL$\301\270\300\303"..., 4096) = 4096
23944 write(249, "\343\2\232Z \257S\334\245y\226R\251\361$F^\326+\333kE\205\307\252J\225Jy\2604\223"..., 2189) = 2189
23944 read(249, 0x7fd880022853, 5)      = -1 EAGAIN (Resource temporarily unavailable)

[...]

23944 read(249, 0x7fd880022853, 5)      = -1 EAGAIN (Resource temporarily unavailable)
23944 read(249,  <unfinished ...>
23944 read(249, "\333\6 \254\16\24\342\341\251\204\203.\355W\345\2\330u}\320jN(\266\377\347.Y>ce\314"..., 80) = 80
23944 write(249, "\27\3\1\0`\7\24V\365\23\255\217\347\317O\332:\0237n\326\236r\373\177\265\315la\26r\247"..., 101) = 101
23944 fsync(249)                        = -1 EINVAL (Invalid argument)
23944 close(249 <unfinished ...>
```

Basically it accepts the connection (`accept`), set a few option on the socket (`fcntl`), starts discussing (`read`, `write`) and when done, closes it (`close`). 
First thing: the `fsync` call is a non sense for socket. I fixed it already in a future PR. 

Now, the communication sometimes goes wrong

```
29286 <... accept resumed> {sa_family=AF_INET6, sin6_port=htons(52592), inet_pton(AF_INET6, "2a07:8500:120:e080:f0ef::d", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 249
29286 getpeername(249,  <unfinished ...>
29286 fcntl(249, F_GETFL)               = 0x2 (flags O_RDWR)
29286 fcntl(249, F_SETFL, O_RDWR|O_NONBLOCK) = 0
23944 read(249, "\26\3\1\0\220", 5)     = 5
23944 read(249, "\1\0\0\214\3\1v\t\213:\224Z\266\202\313:\226}\240\234\177\241.\265f\320tpSE\266\263"..., 144) = 144
23944 write(249, "\26\3\1\0:\2\0\0006\3\1\262\306\212\21\201xf\331\264\337\263\366\363\321\370\33j\2\275\352'"..., 4096) = 4096
23944 write(249, "\343\2\232Z \257S\334\245y\226R\251\361$F^\326+\333kE\205\307\252J\225Jy\2604\223"..., 2189) = 2189
23944 read(249, 0x7fd880022853, 5)      = -1 EAGAIN (Resource temporarily unavailable)

[...]

23944 read(249, 0x7fd880022853, 5)      = -1 EAGAIN (Resource temporarily unavailable)
23944 read(249, 0x7fd880022853, 5)      = -1 ECONNRESET (Connection reset by peer)
23944 write(249, "\27\3\1\3\240\220\36\2746\311\315\263\261AB\33\25\376U~\30\351\315\344i\335\236Q&\221\272\245"..., 933) = -1 EPIPE (Broken pipe)
```

And this socket is never closed. That results in message like 
```
ERROR: Error processing proposal ConnectionError: Error while receiving arguments
ERROR: Invalid action proposal unknown
WARN: Error while handshaking (-1, 'Unexpected EOF')
Exception while reading from peer: (104, 'Connection reset by peer')
```

You can easily check that on your server with the following line:
```
[root@lbvobox102 fd]# for pid in $(lsof | grep "can't" | grep python  | awk {'print $2'} | sort -u); do echo "$(lsof -p $pid | grep -c "can't") $pid $(cat /proc/$pid/cmdline )"; done | sort -n
0 27344 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyBookkeeping/BookkeepingManager/opt/dirac/pro/etc/Bookkeeping_BookkeepingManager.cfg
0 27402 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyResourceStatus/ResourceStatus/opt/dirac/pro/etc/ResourceStatus_ResourceStatus.cfg
0 32402 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyDataManagement/FileCatalog/opt/dirac/pro/etc/DataManagement_FileCatalog.cfg
1 27308 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyAccounting/DataStoreHelper/opt/dirac/pro/etc/Accounting_DataStoreHelper.cfg
1 27356 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyTransformation/TransformationManager/opt/dirac/pro/etc/Transformation_TransformationManager.cfg
9 27461 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyWorkloadManagement/JobStateUpdate/opt/dirac/pro/etc/WorkloadManagement_JobStateUpdate.cfg
377 29286 python/opt/dirac/pro/DIRAC/Core/scripts/dirac-service.pyConfiguration/Server/opt/dirac/pro/etc/Configuration_Server.cfg
```

This works because the leaked socket appears as `can't identify protocol` in `lsof`.
So this script gives you in order:
* the number of leaked socket
* the pid
* the command line of the process

Given the ongoing efforts to get ride of pyGSI and move to M2Crypto, I will not attempt to fix it now, butw will keep it in mind for the future.

I do not have a simple script to reproduce the issue though.

@sfayer for info.

**Comment by fstagni** [comment]
Fixes and in prod.
